### PR TITLE
Fixes and cleanup around JSON collection SQL value conversions

### PR DESCRIPTION
**Pull request by roji**
This does work around SQL logic for converting values out of JSON so they can be further queried in SQL. Originally I thought we may want to add a general API on the type mappings, but I don't think that makes sense at this point. The conversions needed for SQLite are concentrated in a single simple method, and there's no real value in refactoring that out to the type mappings (SQL Server doesn't need any). We can always change this in the future, but for now I think we should leave this as an internal implementation detail of each provider.

The update pipeline doesn't need to perform server-side conversions to JSON; we simply generate the JSON representation we want client-side, and send the appropriate parameter for that (e.g. a JSON string representation of a timestamp). An alternative approach would be to e.g. send a regular DateTime parameter, and then use SQL to convert that to the JSON string representation. But there's no real advantage in that, and it's complex/not always possible. I've done some cleanup/improvement to partial updating in #31232, but I don't think anything else is needed at this point.

(Note that if we ever want to allow a JSON property to be a key/concurrency token, the update pipeline would need to apply the same conversions as the query pipeline (because it would need to appear in the UPDATE WHERE clause)

Specific notes on what's converted and what's not supported:

* On SQL Server, no SQL conversions are needed.
  * When using OPENJSON with WITH (no ordering preservation), it does the conversion job just fine.
  * When using OPENJSON without WITH (because order needs to be preserved), we add SQL CAST() to convert the JSON value out; this works fine in all cases except for varbinary, since CAST() doesn't handle base64. So varbinary is specifically unsupported in scenarios which require ordering preservation.
* On SQLite:
  * We apply special conversions for DateTime (apply `datetime()` to remove the T) and Guid (uppercase)
  * Decimal, binary and DateTimeOffset are not supported, because there doesn't seem to be a good way to convert them out of JSON.

Related to #30727

### CODE REVIEW

File: src/EFCore.Sqlite.Core/Query/Internal/SqliteQueryableMethodTranslatingExpressionVisitor.cs [hunk 1]
```diff
@@ -448,18 +456,36 @@ protected override Expression VisitExtension(Expression expression)
...
-
             // The "standard" JSON timestamp representation is ISO8601, with a T between date and time; but SQLite's representation has
             // no T. Apply a conversion on the value coming out of json_each.
-            SqliteDateTimeTypeMapping => new SqlFunctionExpression(
-                "datetime", new[] { expression }, isNullable, new[] { true }, typeof(DateTime), typeMapping),
-
-            SqliteGuidTypeMapping => new SqlFunctionExpression(
-                "upper", new[] { expression }, isNullable, new[] { true }, typeof(Guid), typeMapping),
+            SqliteDateTimeTypeMapping
+                => new SqlFunctionExpression("datetime", new[] { expression }, isNullable, new[] { true }, typeof(DateTime), typeMapping),
```
Comment (bricelam) [inline src/EFCore.Sqlite.Core/Query/Internal/SqliteQueryableMethodTranslatingExpressionVisitor.cs:469]: Note, `datetime` doesn't preserve fractional seconds. We use this monstrosity instead to get a canonical representation: `rtrim(rtrim(strftime('%Y-%m-%d %H:%M:%f', $value), '0'), '.')`

Comment (roji) [reply src/EFCore.Sqlite.Core/Query/Internal/SqliteQueryableMethodTranslatingExpressionVisitor.cs:469]: Or dear. I'll take a look.

File: src/EFCore.Sqlite.Core/Query/Internal/SqliteQueryableMethodTranslatingExpressionVisitor.cs [hunk 2]
```diff
@@ -222,12 +222,15 @@ protected override QueryableMethodTranslatingExpressionVisitor CreateSubqueryVis
...
-        {
-            // TODO: In any case, we still ned to pass through the type mapping API for doing any conversions (e.g. for datetime, from JSON
-            // ISO8601 to SQLite's format without the T), see #30677. Do this here.
-        }
+        // If we have a collection column, we know the type mapping at this point (as opposed to parameters, whose type mapping will get
+        // inferred later based on usage in SqliteInferredTypeMappingApplier); we should be able to apply any SQL logic needed to convert
+        // the JSON value out to its relational counterpart (e.g. datetime() for timestamps, see ApplyJsonSqlConversion).
+        //
+        // However, doing it here would interfere with pattern matching in e.g. TranslateElementAtOrDefault, where we specifically check
+        // for a bare column being projected out out of the table - if the user composed any operators over the collection, it's no longer
```
Comment (maumar) [inline src/EFCore.Sqlite.Core/Query/Internal/SqliteQueryableMethodTranslatingExpressionVisitor.cs:230]: projected **out out** of the table

File: test/EFCore.SqlServer.FunctionalTests/Query/NonSharedPrimitiveCollectionsQuerySqlServerTest.cs [hunk 3]
```diff
@@ -54,6 +72,10 @@ SELECT COUNT(*)
 """);
     }
 
+    [ConditionalFact]
+    public override Task Array_of_byte()
+        => base.Array_of_byte();
```
Comment (maumar) [inline test/EFCore.SqlServer.FunctionalTests/Query/NonSharedPrimitiveCollectionsQuerySqlServerTest.cs:77]: missing assertsql

Comment (maumar) [inline test/EFCore.SqlServer.FunctionalTests/Query/NonSharedPrimitiveCollectionsQuerySqlServerTest.cs:77]: maybe we need list of test names to exempt from Check_all_tests_overridden?

**Review by maumar — APPROVED** [review]

**Comment by roji** [comment]
Pay special attention to SQLite Array_of_TimeOnly after #31272 is merged (https://github.com/dotnet/efcore/pull/31272/files#r1264439804 after that's merged)

File: test/EFCore.SqlServer.FunctionalTests/Query/NonSharedPrimitiveCollectionsQuerySqlServerTest.cs (hunk 3 above)
Comment (roji) [reply test/EFCore.SqlServer.FunctionalTests/Query/NonSharedPrimitiveCollectionsQuerySqlServerTest.cs:77]: We could, though I think it's not too bad like this...

**Pull request merged**